File: benchmarks/real_comparison/matrix.go
// Matrix multiplication benchmark - Go
package main

import (
	"fmt"
	"time"
)

const N = 512

var A [N][N]float64
var B [N][N]float64
var C [N][N]float64

func matrixMult() {
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			sum := 0.0
			for k := 0; k < N; k++ {
				sum += A[i][k] * B[k][j]
			}
			C[i][j] = sum
		}
	}
}

func main() {
	// Initialize
	for i := 0; i < N; i++ {
		for j := 0; j < N; j++ {
			A[i][j] = float64(i+j) / N
			B[i][j] = float64(i-j) / N
		}
	}

	start := time.Now()

	for iter := 0; iter < 3; iter++ {
		matrixMult()
	}

	elapsed := time.Since(start)
	fmt.Printf("Go: %dx%d matrix mult, Time: %.2f ms (3 runs), C[0][0]=%.4f\n",
		N, N, float64(elapsed.Milliseconds()), C[0][0])
}

File: benchmarks/real_comparison/matrix.rs
// Matrix multiplication benchmark - Rust
use std::time::Instant;

const N: usize = 512;

fn matrix_mult(a: &[[f64; N]; N], b: &[[f64; N]; N], c: &mut [[f64; N]; N]) {
    for i in 0..N {
        for j in 0..N {
            let mut sum = 0.0;
            for k in 0..N {
                sum += a[i][k] * b[k][j];
            }
            c[i][j] = sum;
        }
    }
}

fn main() {
    let mut a = [[0.0f64; N]; N];
    let mut b = [[0.0f64; N]; N];
    let mut c = [[0.0f64; N]; N];

    // Initialize
    for i in 0..N {
        for j in 0..N {
            a[i][j] = (i + j) as f64 / N as f64;
            b[i][j] = (i as i64 - j as i64) as f64 / N as f64;
        }
    }

    let start = Instant::now();

    for _ in 0..3 {
        matrix_mult(&a, &b, &mut c);
    }

    let elapsed = start.elapsed();
    println!(
        "Rust: {}x{} matrix mult, Time: {:.2} ms (3 runs), C[0][0]={:.4}",
        N,
        N,
        elapsed.as_secs_f64() * 1000.0,
        c[0][0]
    );
}
